working/0723-static-extensions/feature-specification.md

```diff
@@ -315,165 +453,482 @@ extension.
 they are creating instances.*
 
+It is a compile-time error to use a constructor declared in an
+extension as the super-initializer of another constructor.
+
+It is a compile-time error to use a constructor declared in an
+extension as a the target of a redirecting generative constructor.
+
+*A generative constructor declared in an extension must be
+redirecting, and hence must eventually bottom out on a non-redirecting
+generative contructor.  However, unlike in the case of a normal
+redirecting constructor, the instance returned by a redirecting
+constructor added in an extension may be more precise than the static
+type given by the return type of the constructor.*
+
 It is a compile-time error if an extension declaration _D_ declares a
 generic constructor whose name is `C` (which includes declarations using
 `C.new`) or `C.name` for some identifier `name`, if _D_ does not have an
 on-declaration, or the name of the on-declaration is not `C`. Note that `C`
 may be an identifier, or an identifier which is prefixed by an import
 prefix.
 
-An extension can declare a constructor which is not generic, that is, it
-does not declare any formal type parameters, and the constructor return
-type does not receive any actual type arguments.
-
-If an extension declaration _D_ named `E` declares a non-generic
-constructor _D1_ and the on-declaration of _D_ is non-generic then _D1_ is
-treated as a generic constructor that declares zero type parameters and
-passes zero actual type arguments to the constructor return type.
-
-*In other words, these constructors get the same treatment as generic
-constructors, except that the type inference step is a no-op.*
+If an extension declaration is generic, the type parameters declared
+by the extension are in scope in any constructors declared in the
+extension (just as with a constructor declared in a generic class).
 
-If an extension declaration _D_ named `E` declares a non-generic
-constructor _D1_ and the on-declaration of _D_ is generic then _D1_ is
-treated as a generic constructor that declares exactly the same type
-parameters as _D_, and it passes exactly the same actual type arguments to
-the constructor return type as the ones that are passed to the
-on-declaration in the on-type.
-
-*For example:*
-
-```dart
-extension E1<X, Y> on C<X, List<Y>, int> {
-  C.name(X x, Iterable<Y> ys): this(x, ys, 14);
-  // The previous line has the same meaning as the next line:
-  C<X, List<Y>, int>.name<X, Y>(X x, Iterable<Y> ys): this(x, ys, 14);
-}
-
-extension E2<X, Y> on D { // D is non-generic.
-  D.new(X x, Y y);
-  // Same as:
-  D.new<X, Y>(X x, Y y);
-}
-```
-
-#### Resolution of a constructor in an extension
+If an extension declaration declares a factory constructor, then the
+downwards context for the purposes of inference of the body of the
+constructor (or for inferring the arguments to the redirectee in the
+case of a redirecting factory constructor) is the on-type of the
```

Comment:
the instantiated on-type of the extension for the invocation.

(Where we find the type arguments to the extension for this invocation by inference if not explicitly given.)